File: Stack/stack.swift

internal final class Stack<T> {

  internal var count: Int {
    get { return self.index }
  }

  internal var isEmpty: Bool {
    get { return self.index == 0 }
  }

  internal var isFull: Bool {
    get { return self.index == self.items.count }
  }

  private var items: Array<T?>
  private var index: Int

  internal init(count: Int) {
    self.items = [T?](repeating: nil, count: count)
    self.index = 0
  }

  internal func peek() throws -> T? {

    guard !self.isEmpty else {
      throw StackError.emptyStack(reason: "Cannot peek at an empty stack!")
    }

    return self.items[self.index - 1]
  }

  internal func pop() throws -> T? {

    guard !self.isEmpty else {
      throw StackError.emptyStack(reason: "Cannot pop from an empty stack!")
    }

    self.index -= 1
    let item: T? = self.items[self.index]
    self.items[self.index] = nil
    return item
  }

  internal func push(item: T?) throws -> Void {

    guard !self.isFull else {
      throw StackError.fullStack(reason: "Cannot push into a full stack!")
    }

    self.items[self.index] = item
    self.index += 1
  }

}

internal enum StackError: Error {
  case emptyStack(reason: String)
  case fullStack(reason: String)
}

File: Stack/stack.js
class Stack {

  constructor(length) {
    this._items = new Array(length);
    this._index = 0;
  }

  get length() {
    return this._index;
  }

  get isEmpty() {
    return this._index === 0;
  }

  get isFull() {
    return this._index === this._items.length;
  }

  peek() {

    if (this.isEmpty) {
      throw new RangeError("Cannot peek at an empty stack!");
    }

    return this._items[this._index - 1];
  }

  pop() {

    if (this.isEmpty) {
      throw new RangeError("Cannot pop from an empty stack!");
    }

    this._index -= 1;
    const item = this._items[this._index];
    this._items[this._index] = undefined;
    return item;
  }

  push(item) {

    if (this.isFull) {
      throw new RangeError("Cannot push into a full stack!");
    }

    this._items[this._index] = item;
    this._index += 1;
  }

}
